### PR TITLE
[MPS] TCCL - Thunderbolt Collective Communication Library

### PR DESCRIPTION
**Pull request by versi379**
# TCCL - Thunderbolt Collective Communication Library

> Authored by @radimurban, implements #189853.

As of now, PyTorch has no MPS-native communication backend, so `torch.distributed` collectives do not run on Apple Silicon. We introduce TCCL to fill that gap: a low-latency communication backend (`c10d::Backend`) that runs collectives on MPS tensors over RDMA on Thunderbolt 5. It implements all necessary collectives to enable parallelism strategies available on the front-end.

<p align="center">

</p>

**Requirements.** macOS 26.2 and Thunderbolt 5 hardware.
## Design

TCCL is an in-tree `c10d::Backend`, functionally inspired by JACCL and structurally modelled around NCCL for PyTorch integration.

The transport layer utilizes RDMA (Remote Direct Memory Access). RDMA bypasses the OS kernel - up to an order of magnitude lower latency and higher bandwidth than TCP.

rdma_vs_tcp_stack

On fp32, TCCL cuts small-message latency ~56× and reaches ~7.6 GB/s per rank of effective bandwidth.

bench_collectives_pub

### Implementation

TCCL is implemented in 3 main files. A few more files wire the distributed-on-MPS path. Tests are covered below.

|             Main files             | Responsibility |
| :--------------------------------: | :------------- |
|    `ProcessGroupTCCL.{hpp,cpp}`    | `c10d::Backend` subclass (`Options`, `TCCLWork`); 12-step RDMA bootstrap ctor, worker `runLoop`, `enqueueCollective` async scaffold, all collective overrides, and the reduce-op / dtype / mesh-vs-ring dispatch. |
|       `TCCLUtils.{hpp,cpp}`        | Transport layer: `dlopen` librdma + dlsym (`TCCLIBVWrapper`), UC queue-pair lifecycle (INIT→RTR→RTS, post/poll) (`TCCLConnection`), registered staging buffers (`TCCLSharedBuffer`), `TCCLDestination`, link-layer subnet check, Store bootstrap. |
| `ProcessGroupTCCLDetail.{hpp,ipp}` | Algorithm layer: reduction-op templates (SUM/PROD/MIN/MAX over float/int/bool, native `__bf16` add); `kChunkSize`; the shared `mesh_exchange`, `ring_step`, and `bidir_ring_step` primitives (depth-2 pipelined); every mesh + ring collective; `wr_id` encoding. |

### Collectives

The full set for every exposed parallelism strategy. Most run in ring and mesh. `gather` and `scatter` are mesh-only.

| Collective                        | Description                                                       | Ring | Mesh |
| --------------------------------- | ---------------------------------------------------------------- | ---- | ---- |
| `allreduce`                       | reductions SUM / AVG / MIN / MAX / PRODUCT (AVG float-only)      | Yes  | Yes  |
| `reduce`                          | rooted reduction, same ops; runs as all-reduce (result on all)   | Yes  | Yes  |
| `broadcast`                       | root → all, byte-copy (any dtype)                                | Yes  | Yes  |
| `allgather` (list)                | gather each rank's shard into separate output tensors, byte-copy | Yes  | Yes  |
| `_allgather_base`                 | single contiguous rank-major output, byte-copy                   | Yes  | Yes  |
| `allgather_into_tensor_coalesced` | the TP / DTensor all-gather virtual                              | Yes  | Yes  |
| `gather`                          | root collects each rank's tensor, byte-copy                      | No   | Yes  |
| `scatter`                         | root sends each rank its slice, byte-copy                        | No   | Yes  |
| `reduce_scatter` (list)           | reduce per-peer input tensors, scatter the result                | Yes  | Yes  |
| `_reduce_scatter_base`            | FSDP eager reduce-scatter (contiguous in/out)                    | Yes  | Yes  |
| `reduce_scatter_tensor_coalesced` | the TP / sequence-parallel reduce-scatter virtual                | Yes  | Yes  |
| `alltoall_base`                   | all-to-all with split sizes (ring: equal split)                  | Yes  | Yes  |
| `send` / `recv`                   | point-to-point                                                   | /    | /    |
| `barrier`                         | Store-based synchronization, no data path                        | /    | /    |

### Algorithm selection

Selection is dynamic, by node count and payload. Benchmarks show allreduce wins with ring even on a mesh at N={3,4}.

collectives_decision_map_poster

- Rule: allreduce goes to ring for bf16 or payload ≥ 1 MB, mesh otherwise. World size ≤ 2 always uses mesh.
- Macs cannot mesh all-to-all beyond N=4, so ring is a complete alternative, not just a large-message optimization.

### Compute-communication overlap

RDMA is a kernel-bypass, CPU-side API. Each collective runs on a dedicated CPU worker: it drives the RDMA, polls completions, and reduces, while the NIC DMAs the payload over Thunderbolt 5.

- The NIC transfers from a small staging buffer, registered once at init and reused every collective.
- Each chunk is memcpy'd into it from the tensor.
- That memcpy's CPU-side source is a zero-copy view over the same unified-memory storage, so the only real copy is the stage.

Overlap comes from the offload. PyTorch MPS Backend exposes a single serial stream, so a collective cannot hide behind compute on a second GPU stream - but the CPU worker runs it while the GPU stays busy. The hand-off is an `MTLSharedEvent`: the main thread records a signal (non-blocking, via commit-and-continue), the worker waits on it. The GPU is never drained.

cc_overlap

- DDP: real overlap. The all-reduce runs on gradient buckets already computed while the GPU proceeds to later buckets - off the critical path.
- TP: each layer's all-reduce feeds the next matmul, so it sits on the critical path. Hiding it needs chunked/async TP, which PyTorch currently gates to CUDA/symmetric-memory. Noted as future work - default TP already brings significant uplift.

## Evaluation

Measured on 4× Mac mini M4 Pro (48/48/64/64 GB), clustered all-to-all over Thunderbolt 5. N<4 and ring runs use a subset of the machines and connections.

### Correctness

319 in-tree cases across three test classes:

| Axis            | Spans |
| --------------- | ----- |
| **Collectives** | `allreduce`; `reduce` (rooted, non-zero root); `broadcast`; `allgather` (list, `_allgather_base`, DTensor/TP coalesced); `gather` / `scatter` (rooted, non-zero root, mesh-only reject on ring); `reduce_scatter` (list, `_reduce_scatter_base` / FSDP, coalesced); `alltoall_base` (equal, uneven `alltoallv`, multi-chunk); `send`/`recv` (chain, directed pair, multi-chunk, non-adjacent); `barrier`. |
| **Reduce ops**  | SUM, AVG, MIN, MAX, PRODUCT |
| **Dtypes**      | float32 / float16 / bfloat16, int8 / int16 / int32 / int64 / uint8, bool (reductions); complex64 (byte-copy collectives) |
| **Algorithms**  | allreduce, all reduce-scatter forms, and alltoall in mesh and ring; broadcast and allgather ring paths via auto-selection |
| **Sizes**       | small tensors plus large payloads across the 512 KB chunk boundary - a 1 MB multi-chunk `send`/`recv`, a 12 MB `allreduce` that trips the ≥ 1 MB ring auto-switch, and ragged odd sizes (N full chunks + 1 element) forced onto the ring |

An out-of-tree superset (which we used for development) adds a full shape sweep, a memory-layout axis (contiguous, sliced, transposed, narrowed, reshaped, chunk-output views with non-zero storage offset), float64, and broader campaigns - ring-vs-mesh sweeps, UC sustained-load stress, raw-RDMA and collective microbenchmarks, and LLM TP/PP scaling on real models.

### Example workloads

FLUX.1-dev, 4 nodes vs one - DDP LoRA fine-tune (training throughput) and TP inference (denoising latency).

flux_speedup

- DDP is communication-light (all-reduce once per step, overlapped with compute), so it scales near-linearly: ~3.8× on 4 nodes, flat across batch size and LoRA rank.
- TP is latency-bound and shards compute, so its win grows with work per step: 1.19× at 512 to 3.04× at 1280 (avg 2.21×). TP also lifts the memory ceiling - 1920 OOMs on one node, runs on four. Note chart is a log scale.

TCCL is otherwise use-case-agnostic - any `torch.distributed` workload with front-end MPS support runs over it. We have run LLMs (Qwen, Llama, ...) in TP and PP inference; TP=4 reaches ~2.3× decode throughput at large batch.

single_vs_tp_linear

The broader point is enablement: sharding runs models that exceed a single node's memory directly in PyTorch on Apple Silicon - previously not an option.

## Running TCCL

TCCL needs three things:

1. RDMA enabled on each Mac
2. a static IP on each Thunderbolt link
3. a few environment variables at launch

The backend reads everything from env vars plus the standard rendezvous.

<details>
<summary><h3>Spelled-out setup for 2-/4-node</h3></summary>

### 1. One-time per node

Enable RDMA on every node (Recovery Mode, then reboot): https://developer.apple.com/documentation/technotes/tn3205-low-latency-communication-with-rdma-over-thunderbolt

### 2. Per boot / node

```
sudo tbtrdmactl unbridge          # makes rdma_enX devices appear
sudo ifconfig bridge0 destroy     # once per node
```

Give each Thunderbolt cable a static IP. Every cable is its own `/30` subnet, and the two ends get different addresses (conventionally `.1` and `.2` - not always `.1`). The only hard rule: both ends of a cable share a non-link-local `/30`. These IPs just bring the RDMA link up; data never leaves the cable.

A node has one interface (and one `/30`) per cable - a 2-node setup has one, a 4-node full mesh has three per node. Which `enX` reaches which peer is just how the cables are plugged in.

### 3. Launch

**Example A - 2 nodes** (`amadeus` = rank 0, `ludwig` = rank 1, cabled `en2`↔`en2`):

```bash
# on amadeus (rank 0):
sudo ifconfig en2 inet 192.168.0.1 netmask 255.255.255.252
MASTER_ADDR=192.168.0.1 MASTER_PORT=29500 RANK=0 WORLD_SIZE=2 \
  TCCL_DEVICE=rdma_en2  python train.py

# on ludwig (rank 1):
sudo ifconfig en2 inet 192.168.0.2 netmask 255.255.255.252    # .2, the other end
MASTER_ADDR=192.168.0.1 MASTER_PORT=29500 RANK=1 WORLD_SIZE=2 \
  TCCL_DEVICE=rdma_en2  python train.py
```

`TCCL_DEVICE` names the RDMA device (omit it if the node exposes exactly one).

**Example B - 4 nodes, full mesh** (`amadeus`, `ludwig`, `johann`, `joseph`). Each Mac's 3 Thunderbolt ports connect it to every other node, so each configures 3 cables (each its own `/30`) and sets `TCCL_PEER_DEVICES` instead of `TCCL_DEVICE`: a comma-separated list of length `WORLD_SIZE` giving the device this rank uses to reach each peer, own slot empty. `MASTER_ADDR` is rank 0's IP on the caller's cable (rank 0's store binds all interfaces).

```bash
# rank 0 - amadeus:  cabled to ludwig/johann/joseph via en2/en3/en4
sudo ifconfig en2 inet 192.168.12.1 netmask 255.255.255.252   # ↔ ludwig
sudo ifconfig en3 inet 192.168.13.1 netmask 255.255.255.252   # ↔ johann
sudo ifconfig en4 inet 192.168.14.1 netmask 255.255.255.252   # ↔ joseph
MASTER_ADDR=192.168.12.1 MASTER_PORT=29500 RANK=0 WORLD_SIZE=4 \
  TCCL_PEER_DEVICES=",rdma_en2,rdma_en3,rdma_en4"  python train.py

# rank 1 - ludwig:
sudo ifconfig en2 inet 192.168.12.2 netmask 255.255.255.252   # ↔ amadeus
sudo ifconfig en4 inet 192.168.23.1 netmask 255.255.255.252   # ↔ johann
sudo ifconfig en3 inet 192.168.24.1 netmask 255.255.255.252   # ↔ joseph
MASTER_ADDR=192.168.12.1 MASTER_PORT=29500 RANK=1 WORLD_SIZE=4 \
  TCCL_PEER_DEVICES="rdma_en2,,rdma_en4,rdma_en3"  python train.py

# rank 2 - johann:
sudo ifconfig en4 inet 192.168.13.2 netmask 255.255.255.252   # ↔ amadeus
sudo ifconfig en2 inet 192.168.23.2 netmask 255.255.255.252   # ↔ ludwig
sudo ifconfig en3 inet 192.168.34.1 netmask 255.255.255.252   # ↔ joseph
MASTER_ADDR=192.168.13.1 MASTER_PORT=29500 RANK=2 WORLD_SIZE=4 \
  TCCL_PEER_DEVICES="rdma_en4,rdma_en2,,rdma_en3"  python train.py

# rank 3 - joseph:
sudo ifconfig en2 inet 192.168.14.2 netmask 255.255.255.252   # ↔ amadeus
sudo ifconfig en4 inet 192.168.24.2 netmask 255.255.255.252   # ↔ ludwig
sudo ifconfig en3 inet 192.168.34.2 netmask 255.255.255.252   # ↔ johann
MASTER_ADDR=192.168.14.1 MASTER_PORT=29500 RANK=3 WORLD_SIZE=4 \
  TCCL_PEER_DEVICES="rdma_en2,rdma_en4,rdma_en3,"  python train.py
```

`MASTER_ADDR` differs per rank - rank 2 reaches rank 0 over the `192.168.13.x` cable, rank 3 over `192.168.14.x`. If the nodes also share a routable network (LAN/IPv6), use one shared `MASTER_ADDR` for all ranks; the per-cable IPs above are what make it work Thunderbolt-only / offline.

### 4. In your PyTorch script

```python
import torch, torch.distributed as dist
dist.init_process_group(backend="tccl", device_id=torch.device("mps", 0))
# then use dist.all_reduce / DDP / TP exactly as with any backend
```

Addressing and the `enX` mapping are per-cable and come from how things are plugged in - a discovery script can generate both the subnet setup and the `TCCL_PEER_DEVICES` rows, but the backend only needs the env vars above.

### 5. Check setup before launching

```python
import torch.distributed as dist
dist.is_tccl_available()                 # built with USE_TCCL=1?
dist.list_tccl_devices()                 # -> ['rdma_en2', 'rdma_en3', 'rdma_en4']
dist.check_tccl_link_layer('rdma_en2')   # raises with the exact fix if the /30 IP is missing
```
</details>

## Future work

- [ ] Multi-wire - split each transfer across N parallel QPs per peer
- [ ] Many frameworks assume distributed PyTorch on MPS is impossible; we hope TCCL changes that and the community builds on it

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroupTCCL.cpp [hunk 1]
```diff
@@ -0,0 +1,1318 @@
...
+ProcessGroupTCCL::TCCLWork::TCCLWork(
+    OpType opType, uint64_t seq, std::vector<at::Tensor> outputs)
+    : Work(/*rank=*/-1, opType), seq_(seq), outputs_(std::move(outputs)) {
+  // Build the result Future (mirrors ProcessGroupGloo::createFutureAsOutput):
+  // carries a List[Tensor]; its device list is the outputs' non-CPU devices so
+  // device-aware consumers (DDP's reducer) chain correctly.
+
+  future_ = c10::make_intrusive<c10::ivalue::Future>(
+      c10::ListType::create(c10::TensorType::get()));
+}
```
[the remark this answers is not in the document]
Comment (radimurban) [reply torch/csrc/distributed/c10d/ProcessGroupTCCL.cpp]: This is a stale comment from an earlier dev device-aware version we abandoned on purpose. The code intentionally builds the Future without a device list. To my understanding, device list exists to ensure  event ordering across streams against pending async GPU work. But that doesn't apply to this backend because:

- MPS is single-queue, there is no other queue to synchronize with
- The collective runs on a CPU worker thread = no GPU work to track at all
- DDP's reducer still chains correctly: when the Future fires the data is already in place

It's also a measured performance decision as the device list added ~140 µs/op of pure overhead that doesn't synchronize anything. We of course tested the correctness of the no-device Future, so correctness is unaffected.

I'll update the comment.

File: torch/distributed/distributed_c10d.py [hunk 2]
```diff
@@ -370,14 +383,17 @@ class _BackendPlugin(NamedTuple):
...
     default_device_backend_map: dict[str, str] = {
         "cpu": GLOO,
         "cuda": NCCL,
         "xpu": XCCL,
-        "mps": GLOO,
+        # TCCL (Thunderbolt RDMA) is the MPS backend when torch is built with
+        # USE_C10D_TCCL. Without it (_TCCL_AVAILABLE == False) this maps to Gloo,
+        # which will raise on MPS tensors.
+        "mps": TCCL if _TCCL_AVAILABLE else GLOO,
     }
```
[the remark this answers is not in the document]
Comment (radimurban) [reply torch/distributed/distributed_c10d.py]: Yes, this is incorrect handling coming from an earlier dev version. Very easy fix. We remove the if and only map `mps` to TCCL backend.

```diff
default_device_backend_map: dict[str, str] = {
    "cpu": GLOO,
    "cuda": NCCL,
    "xpu": XCCL,
-   "mps": TCCL if _TCCL_AVAILABLE else GLOO,
+   "mps": TCCL,
}
```

I'll commit the update.

File: torch/csrc/distributed/c10d/ProcessGroupTCCL.cpp [hunk 3]
```diff
@@ -0,0 +1,1318 @@
...
+// MTLStorageModeShared the allocator exposes a CPU pointer via
+// getSharedBufferPtr; from_blob wraps it as a CPU tensor over the same bytes.
+// Uses the storage base pointer (not data_ptr()) because on MPS data_ptr()
+// folds storage_offset into id<MTLBuffer> arithmetic, yielding an address the
+// allocator can't map; views/chunks (e.g. _allgather_base) have non-zero
+// storage_offset, so a raw data_ptr() lookup would fail there.
+at::Tensor mpsSharedCpuView(const at::Tensor& mpsTensor) {
+  auto* allocator = at::mps::getIMPSAllocator();
+  const void* storage_ptr = mpsTensor.storage().data_ptr().get();
+  auto [cpu_ptr, retain_count] = allocator->getSharedBufferPtr(storage_ptr);
```
[the remark this answers is not in the document]
Comment (radimurban) [reply torch/csrc/distributed/c10d/ProcessGroupTCCL.cpp]: Yes. Though we can just take `.first` of the tuple instead of voiding it, that's cleaner I think. Will include that change:

```diff
-  auto [cpu_ptr, retain_count] = allocator->getSharedBufferPtr(storage_ptr);                                                                                                                                                                          
+  const void* cpu_ptr = allocator->getSharedBufferPtr(storage_ptr).first;  // [cpu_ptr, retain_count] 
```

**Comment by ivanfioravanti** [comment]
Amazing job!!!

**Comment by versi379** [comment]
Hi @d4l3k @fduwjj @kapilsh @kwen2501 any update on this? Thanks :)
